wasmcloud-chat/actors/messages/src/lib.rs

```diff
@@ -1,9 +1,172 @@
+
+const HMK_ORIGIN_CHANNEL: &str = "origin_channel";
+const HMK_ORIGIN_USER: &str = "origin_user";
+const HMK_TIMESTAMP: &str = "timestamp";
+const HMK_MESSAGE_TEXT: &str = "messagetext";
+const HMK_ID: &str = "message_id";
+
+const MSGTYPE_MESSAGE: &str = "MSG";
+const VALID_MESSAGE_TYPES: [&str; 1] = [MSGTYPE_MESSAGE];
 
 #[actor::init]
 fn init() {
-    // Register your message handlers here
-}
\ No newline at end of file
+    messages::Handlers::register_process_message(process_message);
+}
+
+// Processing a message:
+// 1 - validate the message
+// 2 - publish it to the appropriate message broker subject
+// 3 - emit event to event stream
+// 4 - ACK
+fn process_message(message: ChannelMessage) -> HandlerResult<ProcessAck> {
+    let target = MessageTarget::from(&message);
+    if let MessageTarget::Unknown(message) = target {
+        return Err(format!("Unable to select target for message: {}", message).into());
+    }
+
+    Ok(validate_message(&message)
+        .and_then(|_| publish_broker_message(&target, &message))
+        .and_then(|_| emit_stream_event(&target, &message))
+        .map_or_else(
+            |_| ProcessAck::success(&message.message_id),
+            |_| {
+                ProcessAck::fail(
+                    &message.message_id,
+                    "Did not publish and emit to event stream",
+                )
+            },
+        ))
+}
+
+fn validate_message(message: &messages::ChannelMessage) -> HandlerResult<()> {
+    if message.message_id.is_empty() {
+        return Err("No message ID".into());
+    }
+    if message.target_url == format!("wcchat://users/{}", message.origin_user_id) {
+        return Err("Cannot send messages to yourself".into());
+    }
+    if message.message_text.is_empty() {
+        return Err("Cannot send empty messages".into());
+    }
+    if !VALID_MESSAGE_TYPES
+        .to_vec()
+        .contains(&message.message_type.as_str())
+    {
+        return Err("Invalid/unrecognized message type".into());
+    }
+    Ok(())
+}
+
+fn publish_broker_message(
+    target: &MessageTarget,
+    message: &messages::ChannelMessage,
+) -> HandlerResult<()> {
+    let topic = match target {
+        MessageTarget::User(s) => format!("wcc.events.user.{}", s),
+        MessageTarget::Room(s) => format!("wcc.events.room.{}", s),
+        _ => "".to_string(),
+    };
+    let ce = CloudEvent::new_json(
+        &message.message_id,
+        EVENT_TYPE_MESSAGE_PUBLISHED,
+        EVENT_SOURCE,
+        message.created_on,
+        &message,
+    );
+
+    broker::default().publish(topic, "".to_string(), serde_json::to_vec(&ce)?)?;
+    Ok(())
+}
+
+fn emit_stream_event(
+    target: &MessageTarget,
+    message: &messages::ChannelMessage,
+) -> HandlerResult<()> {
+    let message = message.clone();
+
+    let stream_name = match target {
+        MessageTarget::User(s) => format!("user-{}", s),
+        MessageTarget::Room(s) => format!("room-{}", s),
+        _ => "".to_string(),
```

Comment:
Same as the above message comment, just thinking of saving the time publishing a dummy event / message if it wont be used